subprojects/core/src/integTest/groovy/org/gradle/api/invocation/GradleLifecycleSupportedTypesIntegrationTest.groovy

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.invocation
+
+import org.gradle.api.file.ArchiveOperations
+import org.gradle.api.file.FileSystemOperations
+import org.gradle.api.model.ObjectFactory
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.internal.event.ListenerManager
+import org.gradle.process.ExecOperations
+import org.gradle.tooling.provider.model.ToolingModelBuilderRegistry
+import org.slf4j.Logger
+
+import javax.inject.Inject
+import java.nio.charset.Charset
+import java.util.logging.Level
+
+class GradleLifecycleSupportedTypesIntegrationTest extends AbstractIntegrationSpec {
```

Comment:
❓ Are there tests showing an isolated action is incapable of making changes to the (mutable) state it captures? For instance, a test showing two actions capturing the same state but which cannot see changes from one another.

Reply:
Good point. There are no tests to that effect currently, I'll add one.

Reply:
Test added in #28875